### PR TITLE
feat: data importer

### PR DESCRIPTION
**Pull request by vagetablechicken**
Only support bulk load mode now, insert mode needs refactor.
Bulk load importer will send data blocks & binlog first, then send bulk load index.
Bulk load loader(tablet server) use bulk load mgr to manage data receivers for every table, and if failed, the memory is cleaned by mgr too.

data source: only csv and with header, only local files.

### CODE REVIEW

File: src/sdk/sql_router_sdk.i [hunk 1]
```diff
@@ -28,6 +28,13 @@
 
 %shared_ptr(hybridse::sdk::ResultSet);
 %shared_ptr(hybridse::sdk::Schema);
+%shared_ptr(fedb::sdk::SQLRouter);
```
Comment (dl239) [inline src/sdk/sql_router_sdk.i]: `fedb` has renamed `openmldb`. Remove these unused code

File: src/storage/segment.cc [hunk 2]
```diff
@@ -189,6 +189,52 @@ void Segment::Put(const Slice& key, uint64_t time, DataBlock* row) {
     idx_byte_size_.fetch_add(byte_size, std::memory_order_relaxed);
 }
 
+void Segment::BulkLoadPut(int key_entry_id, const Slice& key, uint64_t time, DataBlock* row) {
+    void* key_entry_or_list = nullptr;
+    uint32_t byte_size = 0;
+    std::lock_guard<std::mutex> lock(mu_);  // TODO(hw): need lock?
+    int ret = entries_->Get(key, key_entry_or_list);
+    if (ts_cnt_ == 1) {
```
Comment (dl239) [inline src/storage/segment.cc:201]: use one method with `Put(const Slice& key, uint64_t time, DataBlock* row)`

File: src/tablet/tablet_impl.cc [hunk 3]
```diff
@@ -4766,5 +4769,140 @@ void TabletImpl::CreateProcedure(const std::shared_ptr<hybridse::sdk::ProcedureI
...
+
+    if (follower_.load(std::memory_order_relaxed)) {
+        response->set_code(::openmldb::base::ReturnCode::kIsFollowerCluster);
+        response->set_msg("is follower cluster");
+        return;
+    }
+    uint64_t start_time = ::baidu::common::timer::get_micros();
+    auto tid = request->tid();
+    auto pid = request->pid();
+    // TODO(hw): tid-pid use xx-xx, 方便搜索
```
Comment (dl239) [inline src/tablet/tablet_impl.cc]: use english comment

File: src/tablet/bulk_load_mgr.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * limitations under the License.
+ */
+
+#ifndef SRC_TABLET_BULK_LOAD_MGR_H_
+#define SRC_TABLET_BULK_LOAD_MGR_H_
+
+#include <map>
+#include <memory>
+#include <mutex>
+
```
Comment (dl239) [inline src/tablet/bulk_load_mgr.h:22]: resolve this cpplint error

File: src/tablet/bulk_load_mgr.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#include "replica/log_replicator.h"
+#include "storage/mem_table.h"
+#include "tablet/data_receiver.h"
+
+namespace openmldb::tablet {
+class BulkLoadMgr {
+ public:
+    bool DataAppend(uint32_t tid, uint32_t pid, const ::openmldb::api::BulkLoadRequest* request,
```
Comment (dl239) [inline src/tablet/bulk_load_mgr.h]: rename to `AppendData`

File: src/tablet/data_receiver.cc [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        auto buf = new char[info.length()];  // TODO(hw): use pool
+        iter.copy_and_forward(buf, info.length());
+        data_blocks_.emplace_back(new storage::DataBlock(info.ref_cnt(), buf, info.length(), true));
+    }
+    if (iter.bytes_left() != 0) {
+        LOG(ERROR) << tid_ << "-" << pid_ << " data and info mismatch, revert this part";
+        // Not only ptr, DataBlock needs delete.
+        for (auto block_iter = data_blocks_.begin() + last_block_size; block_iter != data_blocks_.end();) {
+            // regardless of dim_cnt_down
+            delete *block_iter;
```
Comment (dl239) [inline src/tablet/data_receiver.cc]: use range erase after loops

File: src/tablet/data_receiver.cc [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+    if (!request->has_part_id() || !PartValidation(request->part_id())) {
+        LOG(WARNING) << tid_ << "-" << pid_ << " data receiver received invalid part id, expect " << next_part_id_
+                     << ", actual "
+                     << (request->has_part_id() ? "no id" : std::to_string(request->part_id()));
+        return false;
+    }
+
+    // We must copy data from IOBuf, cuz the rows have different TTLs, it's not a good idea to keep them together.
+    butil::IOBufBytesIterator iter(data);
```
Comment (dl239) [inline src/tablet/data_receiver.cc:32]: validation need not lock, move lock here

File: src/tablet/data_receiver.cc [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                        if (block == nullptr) {
+                            // TODO(hw): valid?
+                            continue;
+                        }
+                        ::openmldb::api::LogEntry entry;
+                        entry.set_pk(pk);
+                        entry.set_ts(time_entry.time());
+                        entry.set_value(block->data, block->size);
+                        entry.set_term(replicator->GetLeaderTerm());
+                        replicator->AppendEntry(entry);
```
Comment (dl239) [inline src/tablet/data_receiver.cc]: has lost ts information when index has multi ts cols

File: src/storage/mem_table.cc [hunk 9]
```diff
@@ -739,6 +740,98 @@ TableIterator* MemTable::NewTraverseIterator(uint32_t index) {
     return new MemTableTraverseIterator(segments_[real_idx], seg_cnt_, ttl->ttl_type, expire_time, expire_cnt, 0);
 }
 
+bool MemTable::GetBulkLoadInfo(::openmldb::api::BulkLoadInfoResponse* response) {
+    response->set_seg_cnt(seg_cnt_);
+
+    // TODO(hw): out of range will get -1, only a temporary solution.
+    uint32_t idx = 0;
+    int32_t pos;
+    while ((pos = table_index_.GetInnerIndexPos(idx)) != -1) {
```
Comment (dl239) [inline src/storage/mem_table.cc:749]: why not set `add_inner_index_pos` in next allinnerindex loop

File: java/openmldb-import/pom.xml [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+          <groupId>com.4paradigm.openmldb</groupId>
+          <artifactId>openmldb-jdbc</artifactId>
+          <version>0.2.0</version>
+        </dependency>
+        <dependency>
+          <groupId>com.4paradigm.openmldb</groupId>
+          <artifactId>openmldb-native</artifactId>
```
Comment (dl239) [inline java/openmldb-import/pom.xml]: need not add `openmldb-native` dependency explicitly

File: java/openmldb-import/src/main/java/com/_4paradigm/openmldb/importer/RowCodecCommon.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                        && column == null) {
+                    throw new RuntimeException("col " + columnDesc.getName() + " should not be null");
+                }
+                strLength += ((String) column).length();
+            }
+        }
+        return strLength;
+    }
+
+//    public static String getPrimaryKey(Map<String, Object> row, List<ColumnKey> columnKeyList, List<ColumnDesc> schema) {
```
Comment (dl239) [inline java/openmldb-import/src/main/java/com/_4paradigm/openmldb/importer/RowCodecCommon.java]: remove unused code

File: java/openmldb-import/src/main/java/com/_4paradigm/openmldb/importer/InsertImporter.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+            return;
+        }
+        Schema schema = insertRow.GetSchema();
+
+        // TODO(hw): record.getParser().getHeaderMap() check schema? In future, we may read multi files, so check schema in each worker.
+        // What if the header of record is missing?
+        List<String> stringCols = getStringColumnsFromSchema(schema);
+        for (int i = range.getLeft(); i < range.getRight(); i++) {
+            // insert placeholder
+            SQLInsertRow row = router.getInsertRow(dbName, insertPlaceHolder);
```
Comment (dl239) [inline java/openmldb-import/src/main/java/com/_4paradigm/openmldb/importer/InsertImporter.java:75]: SQLInsertRow is c++ object, need delete explicitly after used?

File: src/tablet/data_receiver.cc (hunk 8 above)
Comment (vagetablechicken) [reply src/tablet/data_receiver.cc]: It's calculated by importer IndexRegionBuilder. Won't lost. Each entry in index is only one entry(may be one of multi ts)

File: src/storage/mem_table.cc (hunk 9 above)
Comment (vagetablechicken) [reply src/storage/mem_table.cc:749]: This one is to get `column_key_2_inner_index_`, the next loop is to get `inner_indexs_`. The names are too easy to get confused.

File: java/openmldb-import/src/main/java/com/_4paradigm/openmldb/importer/InsertImporter.java (hunk 12 above)
Comment (vagetablechicken) [reply java/openmldb-import/src/main/java/com/_4paradigm/openmldb/importer/InsertImporter.java:75]: java own it. `this(sql_router_sdkJNI.new_SQLInsertRow(..., true);`.
The last arg is `swigCMemOwn`.
finalize() will delete.

File: src/tablet/data_receiver.cc (hunk 8 above)
Comment (vagetablechicken) [reply src/tablet/data_receiver.cc:113]: It's calculated by importer IndexRegionBuilder. Won't lost. Each entry in index is only one entry(may be one of multi ts)

File: src/tablet/data_receiver.cc (hunk 6 above)
Comment (vagetablechicken) [reply src/tablet/data_receiver.cc]: use resize

File: src/storage/mem_table.cc (hunk 9 above)
Comment (dl239) [reply src/storage/mem_table.cc:749]: yes, `GetInnerIndex` and `GetInnerIndexPos`  is not clear

File: src/tablet/data_receiver.cc (hunk 8 above)
Comment (vagetablechicken) [reply src/tablet/data_receiver.cc]: binlog needs a largely modifying, working on it.

File: src/tablet/tablet_impl.cc [hunk 13]
```diff
@@ -4766,5 +4769,147 @@ void TabletImpl::CreateProcedure(const std::shared_ptr<hybridse::sdk::ProcedureI
...
+            replicator = GetReplicator(request->tid(), request->pid());
+            if (!replicator) {
+                PDLOG(WARNING, "fail to find table tid %u pid %u leader's log replicator", request->tid(),
+                      request->pid());
+                break;
+            }
+
+            auto ok = bulk_load_mgr_.WriteBinlogToReplicator(tid, pid, replicator, request);
+            if (!ok) {
+                DLOG(INFO) << "write binlog failed";
```
Comment (dl239) [inline src/tablet/tablet_impl.cc]: update to `LOG(WARNING)`. add tid pid info

File: src/tablet/data_receiver.cc [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        if (info.ts_dimensions_size() > 0) {
+            entry.mutable_ts_dimensions()->CopyFrom(info.ts_dimensions());
+        }
+        entry.set_ts(info.time());
+        replicator->AppendEntry(entry);
+    }
+    LOG(INFO) << "binlog write num " << request->binlog_info_size();
+    if (FLAGS_binlog_notify_on_put) {
+        replicator->Notify();
```
Comment (dl239) [inline src/tablet/data_receiver.cc]: has notified in tablet/tablet_impl.cc 4885

File: src/tablet/data_receiver.cc [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+            << "WriteBinlogToReplicator follows AppendData, the same request, the same part id, but cur part id"
+            << next_part_id_ - 1 << ", request part id " << request->part_id();
+        return false;
+    }
+    for (int i = 0; i < request->binlog_info_size(); ++i) {
+        const auto& info = request->binlog_info(i);
+        ::openmldb::api::LogEntry entry;
+        auto* block = info.block_id() < data_blocks_.size() ? data_blocks_[info.block_id()] : nullptr;
+        if (block == nullptr) {
+            LOG(ERROR) << "binlog wants " << info.block_id() << ", but cached block size = " << data_blocks_.size();
```
Comment (dl239) [inline src/tablet/data_receiver.cc:104]: WARNING

Comment (vagetablechicken) [reply src/tablet/data_receiver.cc:104]: It's an error, not a warning which can be ignored

File: java/openmldb-import/src/main/resources/log4j.xml [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        <level value="warn"/>
+        <appender-ref ref="myConsole"/>
+    </logger>
+
+    <!-- 根logger的设置-->
+    <root>
+        <priority value="warn"/>
+        <appender-ref ref="myConsole"/>
+    </root>
+</log4j:configuration>
```
Comment (vagetablechicken) [inline java/openmldb-import/src/main/resources/log4j.xml]: - [ ]

File: src/tablet/tablet_impl_test.cc [hunk 17]
```diff
@@ -5593,7 +5593,81 @@ TEST_F(TabletImplTest, SendIndexData) {
...
+        request.add_index_region();
+        ::fedb::api::GeneralResponse response;
+        MockClosure closure;
+        brpc::Controller cntl;
+        cntl.request_attachment().append("123");
+        tablet.BulkLoad(&cntl, &request, &response, &closure);
+        ASSERT_EQ(0, response.code()) << response.msg();
+    }
+
+    // get data from the table
```
Comment (vagetablechicken) [inline src/tablet/tablet_impl_test.cc]: - [ ]

File: src/tablet/data_receiver.cc (hunk 7 above)
Comment (vagetablechicken) [reply src/tablet/data_receiver.cc:32]: Validation's part id may be changed by any method which will do part check. Keep the simple logic.

**Review by dl239 — APPROVED** [review]

**Review by imotai — APPROVED** [review]

**Pull request merged**